gpflow/likelihoods/likelihoods.py

```diff
@@ -69,92 +69,307 @@ def inv_probit(x):
+    def conditional_variance(self, F):
         """
+        The conditional marginal variance of Y|F: [var(Y₁|F), ..., var(Yₖ|F)]
+        where K = observation_dim
+        :param F: function evaluation Tensor, with shape [..., latent_dim]
+        :return variance [..., observation_dim]
+        """
+        self.check_latent_dims(F)
+        var_Y = self._conditional_variance(F)
+        self.check_data_dims(var_Y)
+        return var_Y
 
-        def integrand(*X):
-            return self.conditional_variance(*X) + self.conditional_mean(*X) ** 2
+    def _conditional_variance(self, F):
+        raise NotImplementedError
 
-        integrands = [self.conditional_mean, integrand]
-        nghp = self.num_gauss_hermite_points
-        E_y, E_y2 = ndiagquad(integrands, nghp, Fmu, Fvar)
-        V_y = E_y2 - E_y ** 2
-        return E_y, V_y
+    def predict_mean_and_var(self, Fmu, Fvar):
+        """
+        The conditional mean and marginal variance of Yₖ|F
+        :param Fmu: mean function evaluation Tensor, with shape [..., latent_dim]
+        :param Fvar: variance of function evaluation Tensor, with shape [..., latent_dim]
+        :return mean and variance, both with shape [..., observation_dim]
+        """
+        self.check_latent_dims(Fmu)
+        self.check_latent_dims(Fvar)
+        mu, var = self._predict_mean_and_var(Fmu, Fvar)
+        self.check_data_dims(mu)
+        self.check_data_dims(var)
+        return mu, var
+
+    def _predict_mean_and_var(self, Fmu, Fvar):
+        raise NotImplementedError
 
     def predict_density(self, Fmu, Fvar, Y):
         r"""
         Given a Normal distribution for the latent function, and a datum Y,
         compute the log predictive density of Y.
 
         i.e. if
-            q(f) = N(Fmu, Fvar)
+            q(F) = N(Fmu, Fvar)
 
         and this object represents
 
-            p(y|f)
+            p(y|F)
 
         then this method computes the predictive density
 
-            \log \int p(y=Y|f)q(f) df
+            log ∫ p(y=Y|F)q(F) df
 
-        Here, we implement a default Gauss-Hermite quadrature routine, but some
-        likelihoods (Gaussian, Poisson) will implement specific cases.
+        :param Fmu: mean function evaluation Tensor, with shape [..., latent_dim]
+        :param Fvar: variance of function evaluation Tensor, with shape [..., latent_dim]
+        :param Y: observation Tensor, with shape [..., observation_dim]:
+        :return predicted density, with shape [...]
         """
-        integrand = self.log_prob
-        nghp = self.num_gauss_hermite_points
-        return ndiagquad(integrand, nghp, Fmu, Fvar, logspace=True, Y=Y)
+        tf.debugging.assert_equal(tf.shape(Fmu), tf.shape(Fvar))
+        self._check_last_dims_valid(Fmu, Y)
+        res = self._predict_density(Fmu, Fvar, Y)
+        self.check_return_shape(res, Fmu, Y)
+        return res
+
+    def _predict_density(self, Fmu, Fvar, Y):
```

Comment:
```suggestion
    @abc.abstractmethod
    def _predict_density(self, Fmu, Fvar, Y):
```

Reply:
Actually, could we call the "implementation" `_predict_log_density` already? I think that'd be more accurate. And rename `predict_density` to `predict_log_density` ? (We can add a `predict_density()` function that calls `predict_log_density` and throws a DeprecationWarning)

Reply:
Maybe changing the interface (`predict_density` -> `predict_log_density`) should be a separate PR, just to make it easier to review, but change `_predict_density` to `_predict_log_density` in this PR already? would you be up for doing the second PR? I'd be happy to review. @jameshensman @thevincentadam

Reply:
just did it anyway, not that big a change